zebra-consensus/src/block.rs

```diff
@@ -177,15 +178,15 @@ where
 
             let known_utxos = new_outputs(&block, &transaction_hashes);
             for transaction in &block.transactions {
-                let rsp = transaction_verifier
-                    .ready_and()
-                    .await
-                    .expect("transaction verifier is always ready")
-                    .call(tx::Request::Block {
-                        transaction: transaction.clone(),
-                        known_utxos: known_utxos.clone(),
-                        height,
-                    });
+                let request = tx::Request::Block {
+                    transaction: transaction.clone(),
+                    known_utxos: known_utxos.clone(),
+                    height,
+                };
+
+                // We use a `ServiceExt::oneshot`, so that transaction_verifier
+                // service `poll_ready` has a corresponding `call`. See #1593.
+                let rsp = transaction_verifier.clone().oneshot(request);
```

Comment:
oneshot takes ownership of the service, but rustc can't be sure that the transaction_verifier is Send:
https://doc.rust-lang.org/std/marker/trait.Send.html

Send means that the type can be safely sent between threads. Previously, the code was just sending a mutable reference to the verifier between threads, which uses the Sync bound:
https://doc.rust-lang.org/std/marker/trait.Sync.html

Since futures can be sent between threads, the transaction verifier needs to be wrapped in a Buffer, which is a Send type:
https://github.com/ZcashFoundation/zebra/pull/1700/files/140b83b4bab1a161a2bc2a3a3c39e77136687dd6#diff-49f13a370043562a8c8c03657175ad9a6a49f098776eb88e8bfc08d02a1be148

We are going to need a Buffered transaction verifier when we add a mempool anyway. With a shared transaction verifier, we can verify block transactions and loose transactions using the same signature and proof verifiers, to take advantage of batched cryptography.

Reply:
You should probably also drop the Derive Clone on the transaction and script verifiers. We want to use the buffer instead of cloning the underlying verifier.

I think it's ok to just add the buffer in BlockVerifier::new, but if there are any other transaction verifier users, we should add the buffer in an init method in transaction.rs. (Similarly for the script verifier, but I'm not sure if it will ever do batching.)

Reply:
@yaahc @dconnolly will the script verifier ever do batch verification?

I'm trying to work out if it's worth wrapping it in a Buffer.
(This isn't an urgent question, I'd just like to know.)